### PR TITLE
feat(*): add breadcrumb component react

### PR DESCRIPTION
**Pull request by stif59100**
ref:MANAGER-12702

| Question         | Answer
| ---------------- | ---
| Branch?          | `feat/MANAGER-12701
| Bug fix?         | no
| New feature?     | yes
| Breaking change? | no
| Tickets          | MANAGER-12702
| License          | BSD 3-Clause

- [x] Try to keep pull requests small so they can be easily reviewed.
- [x] Commits are signed-off
- [ ] ~~Only FR translations have been updated~~
- [x] Branch is up-to-date with target branch
- [x] Lint has passed locally
- [x] Standalone app was ran and tested locally
- [x] Ticket reference is mentioned in linked commits (internal only)
- [ ] ~~Breaking change is mentioned in relevant commits~~

## Description

I added breadcrumb component react

### CODE REVIEW

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+type MatchHandle = {
+  breadcrumb?: useBreadcrumbHandle;
+};
+
+export interface useBreadcrumbProps {
+  rootLabel?: string;
+}
+
+export const useBreadcrumb = ({ rootLabel }: useBreadcrumbProps) => {
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:31]: It's a component, not a hook

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  const isMatchHandle = (obj: any): obj is MatchHandle => {
+    return obj && typeof obj === 'object' && 'breadcrumb' in obj;
+  };
+
+  const crumbs = [...root, ...matchCrumbs].filter(
+    (crumb) => crumb.label !== undefined && crumb.label !== '',
+  );
+
+  const data = crumbs.map((crumb) => ({
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:103]:
```suggestion
  const data = [...root, ...matchCrumbs].filter(
    (crumb) => crumb.label !== undefined && crumb.label !== '',
  ).map((crumb) => ({
```

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+              };
+              return crumb;
+            }
+            return {
+              label: match.params.serviceName,
+              href: match.pathname,
+            };
+          }),
+        );
+        console.log(items);
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:82]:
```suggestion

```

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+  useEffect(() => {
+    const fetchMatchCrumbs = async () => {
+      try {
+        const items = await Promise.all(
+          matches.map(async (match) => {
+            const { handle, data, params } = match;
+
+            if (isMatchHandle(handle)) {
+              const breadcrumb = await Promise.resolve(handle.breadcrumb);
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:68]:
```suggestion
              const { breadcrumb } = handle;
```

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+              const label = await breadcrumb?.({ data, params });
+              const crumb = {
+                label: label ?? '',
+                href: match.pathname,
+              };
+              return crumb;
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:74]:
```suggestion
              return {
                label: label ?? '',
                href: match.pathname,
              };
```

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  const matches = useMatches();
+  const shell = useShell();
+  const [root, setRoot] = useState<useBreadcrumbItem[]>([]);
+  const [matchCrumbs, setMatchCrumbs] = useState<useBreadcrumbItem[]>([]);
+
+  useEffect(() => {
+    const fetchRoot = async () => {
+      try {
+        const response = await shell.navigation.getURL('hub', '#/', {});
+        console.log(response);
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:41]:
```suggestion

```

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        if (typeof response === 'string') {
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:42]: what are the other possible types for response ?
If it's just undefined or null, no need to use `typeof` : `if (!!response) {`

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx (hunk 4 above)
Comment (chipp972) [reply packages/manager/core/application/src/hooks/useBreadcrumb.tsx:59]: I don't understand where the "basic crumb" is called on this line

File: packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs [hunk 8]
```diff
@@ -1,67 +1,21 @@
...
+
+function Breadcrumb({ rootLabel }: BreadcrumbProps): JSX.Element {
+  const breadcrumbItems = useBreadcrumb({ rootLabel });
+
+  return (
+    <OsdsBreadcrumb
+      items={breadcrumbItems.map((item) => ({
+        label: item.label,
+        href: item.href,
+      }))}
```
Comment (chipp972) [inline packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs:16]: Do we really need this mapping ?

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+  const isMatchHandle = (obj: any): obj is MatchHandle => {
+    return obj && typeof obj === 'object' && 'breadcrumb' in obj;
+  };
+
+  const crumbs = [...root, ...matchCrumbs].filter(
+    (crumb) => crumb.label !== undefined && crumb.label !== '',
+  );
+
+  return crumbs;
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:93]:
```suggestion
  return [...root, ...matchCrumbs].filter(
    (crumb) => crumb.label !== undefined && crumb.label !== '',
  );
```

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      }
+    };
+
+    fetchMatchCrumbs();
+  }, [matches]);
+
+  const isMatchHandle = (obj: any): obj is MatchHandle => {
+    return obj && typeof obj === 'object' && 'breadcrumb' in obj;
+  };
+
```
Comment (chipp972) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:88]: Define this outside the hook

File: packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs [hunk 11]
```diff
@@ -1,67 +1,21 @@
-/**
- *  TODO - Move in the future
- *  Super components manager library
- */
-
-import React, { useEffect, useState } from 'react';
-import { Params, useMatches } from 'react-router-dom';
+import React from 'react';
```
Comment (chipp972) [inline packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs:1]: We should add this component to the manager-components, shouldn't we ?

File: packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs (hunk 8 above)
Comment (chipp972) [reply packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs:16]: But breadcrumbItem is already an array and items already has props label and href. So the only thing interesting here is to filter other properties but I don't think there are other properties that could impact the OsdsBreadcrumb

File: packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs (hunk 11 above)
Comment (chipp972) [reply packages/manager/core/generator/app/templates/src/components/Breadcrumb/Breadcrumb.tsx.hbs:1]: Ok :+1:

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx (hunk 4 above)
Comment (chipp972) [reply packages/manager/core/application/src/hooks/useBreadcrumb.tsx:59]: if handle.breadcrumb is potentially a promise, we should just `await handle.breadcrumb`
`await Promise.resolve(x)` is equivalent to `x`

**Review by chipp972 — DISMISSED** [review]

**Review by aboungnaseng-ovhcloud — APPROVED** [review]

**Review by chipp972 — DISMISSED** [review]

**Review by aboungnaseng-ovhcloud — DISMISSED** [review]

File: packages/manager/apps/catalog/src/pages/index.tsx [hunk 12]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (anooparveti) [inline packages/manager/apps/catalog/src/pages/index.tsx:1]: **question:** Why disabling the Lint errors.

File: packages/manager/core/application/src/hooks/useBreadcrumb.tsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+export const useBreadcrumb = ({ rootLabel }: BreadcrumbProps) => {
+  const matches = useMatches();
+  const shell = useShell();
+  const [root, setRoot] = useState<BreadcrumbItem[]>([]);
+  const [matchCrumbs, setMatchCrumbs] = useState<BreadcrumbItem[]>([]);
+  useEffect(() => {
+    const fetchRoot = async () => {
+      try {
+        const response = await shell.navigation.getURL('hub', '#/', {});
```
Comment (cbourgois) [inline packages/manager/core/application/src/hooks/useBreadcrumb.tsx:37]: Root url is always `hub` app url ?

File: packages/manager/apps/catalog/src/components/Breadcrumb/Breadcrumb.tsx [hunk 14]
```diff
@@ -1,74 +1,11 @@
...
-        const validCrumbs = breadcrumbs.filter(
-          ({ label, href }) => !!label && !!href,
-        );
-        setMatchCrumbs(validCrumbs);
-      })
-      .catch((error) => {
-        console.error('Error fetching breadcrumbs:', error);
-      });
-  }, [matches]);
+  const breadcrumbItems = useBreadcrumb({ rootLabel: 'Dashboard' });
```
Comment (frenautvh) [inline packages/manager/apps/catalog/src/components/Breadcrumb/Breadcrumb.tsx:6]: should this label be translated ?

**Review by aboungnaseng-ovhcloud — APPROVED** [review]

**Review by anooparveti — APPROVED** [review]

**Pull request merged**